### PR TITLE
using cv2 dnn interface to run yolov8 model

### PR DESCRIPTION
**Pull request by Tsai-chia-hsiang**
This is a sample code for using opencv dnn interface to run ultralytics yolov8 model for object detection.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [X] I agree to contribute to the project under Apache 2 License.
- [X] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [X] The PR is proposed to the proper branch
- [] There is a reference to the original bug report and related work
- [] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: samples/dnn/dnn_model_runner/dnn_conversion/ultralytics/predict/dog1.jpg
Comment (dkurt) [inline samples/dnn/dnn_model_runner/dnn_conversion/ultralytics/predict/dog1.jpg:1]: Do not add images

File: samples/dnn/dnn_model_runner/dnn_conversion/ultralytics/yolov8_cv.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    [189, 214, 111], [214, 230, 107], [81, 248, 106], 
+    [37, 94, 186], [116, 15, 188], [66, 180, 87], 
+    [171, 200, 14], [205, 103, 83], [221, 92, 177], 
+    [93, 137, 220], [18, 79, 251], [88, 243, 155], 
+    [136, 208, 232], [121, 115, 245], [12, 213, 244], 
+    [174, 155, 25], [42, 47, 82], [56, 28, 90], [103, 33, 105]
+]
+
+
+class Opencv_Yolov8:
```
Comment (dkurt) [inline samples/dnn/dnn_model_runner/dnn_conversion/ultralytics/yolov8_cv.py:41]: There is https://github.com/opencv/opencv/blob/4.x/samples/dnn/object_detection.py. Can it be used?

File: samples/dnn/dnn_model_runner/dnn_conversion/ultralytics/predict/dog1.jpg
Comment (Tsai-chia-hsiang) [reply samples/dnn/dnn_model_runner/dnn_conversion/ultralytics/predict/dog1.jpg:1]: Hi, I have deleted the images, please check, thanks!

File: samples/dnn/dnn_model_runner/dnn_conversion/ultralytics/yolov8_cv.py (hunk 1 above)
Comment (Tsai-chia-hsiang) [reply samples/dnn/dnn_model_runner/dnn_conversion/ultralytics/yolov8_cv.py:47]: Yes. But if you want to make predictions quickly, Yolo model is a good option.  This sample code seems to use only fast-RCNN and SSD model, both of they are slower than Yolov8, even though they are good model.

Comment (dkurt) [reply samples/dnn/dnn_model_runner/dnn_conversion/ultralytics/yolov8_cv.py:47]: The sample supports YOLO models, take a look closer at https://github.com/opencv/opencv/blob/0507043a55e49069bf71b4ec629367237e03b723/samples/dnn/object_detection.py#L148

Comment (dkurt) [reply samples/dnn/dnn_model_runner/dnn_conversion/ultralytics/yolov8_cv.py:47]: Good question. Code sample worked only with Yolo from Darknet, not ONNX. Can you please show output layers of YOLOv8 using https://netron.app/?

Comment (Tsai-chia-hsiang) [reply samples/dnn/dnn_model_runner/dnn_conversion/ultralytics/yolov8_cv.py:47]: Here is the visualization of netron for yolov8 : 

yolov8

output layer: float32[1,84,8400]

Please check, thanks!

Comment (dkurt) [reply samples/dnn/dnn_model_runner/dnn_conversion/ultralytics/yolov8_cv.py:47]: Thanks! So, there is no a unique layer type by which OpenCV can determine YOLO.

Can you please still try to fit your sample into https://github.com/opencv/opencv/blob/4.x/samples/dnn/object_detection.py but add an argument to run YOLO postprocessing?

Comment (Bleach665) [reply samples/dnn/dnn_model_runner/dnn_conversion/ultralytics/yolov8_cv.py:47]: I understand that it is not a good idea but you can determine Yolo by check output Mat `size`.
Result is `1 x 25200 x 85` for Yolov5 and `1 x 84 x 8400` for Yolov8.

Comment (Tsai-chia-hsiang) [reply samples/dnn/dnn_model_runner/dnn_conversion/ultralytics/yolov8_cv.py]: @dkurt 
Hi, we have been committed the refactorized ```object_detection.py``` and ```model.yaml``` to make it enable 
to run yolov8 model under ONNX framework. 

Please check them, thanks.

File: samples/dnn/object_detection.py [hunk 2]
```diff
@@ -62,31 +62,40 @@
...
-# Load names of classes
-classes = None
-if args.classes:
-    with open(args.classes, 'rt') as f:
-        classes = f.read().rstrip('\n').split('\n')
-
-# Load a network
-net = cv.dnn.readNet(cv.samples.findFile(args.model), cv.samples.findFile(args.config), args.framework)
+model_length = len(args.model)
+if args.model[model_length-5:model_length] != '.onnx':
```
Comment (dkurt) [inline samples/dnn/object_detection.py:66]: Why we need so? Please just add an argument like `--postprocessing` with values "yolo", "ssd" and others

File: samples/dnn/object_detection.py [hunk 3]
```diff
@@ -62,31 +64,40 @@
...
+    # Load a network
+    net = cv.dnn.readNet(cv.samples.findFile(args.model), cv.samples.findFile(args.config), args.framework)
+# onnx_yolov8
+else:
+    classes = None
+    if args.classes:
+        with open(args.classes, 'rt') as f:
+            classes = f.read().rstrip('\n').split('\n')
+    net = cv.dnn.readNetFromONNX(args.model)
+###
```
Comment (dkurt) [inline samples/dnn/object_detection.py:100]: Why we need a separate method `readNetFromONNX`? Can `readNet` cover onnx models?

File: samples/dnn/object_detection.py [hunk 4]
```diff
@@ -164,6 +175,24 @@ def drawPred(classId, conf, left, top, right, bottom):
                     classIds.append(classId)
                     confidences.append(float(confidence))
                     boxes.append([left, top, width, height])
+    # yolov8
+    elif lastLayer.type == 'Identity':
```
Comment (dkurt) [inline samples/dnn/object_detection.py:179]: Check for args.postprocessing

File: samples/dnn/object_detection.py (hunk 3 above)
Comment (JasonLin1110) [reply samples/dnn/object_detection.py:100]: ```readNet``` can cover onnx models， but we still needs special handling for yolov8 as it doesn't require ```args.cofig``` and ```args.framework```, we will change code to ```net = cv.dnn.readNet(args.model)```

File: samples/dnn/object_detection.py [hunk 5]
```diff
@@ -263,6 +310,8 @@ def processingThreadBody():
 
 
         if not frame is None:
+            if args.postprocessing == 'yolov8':
+                frame = cv.resize(frame, (args.height, args.width))
```
[the remark this answers is not in the document]
Comment (JasonLin1110) [reply samples/dnn/object_detection.py:314]: In YOLOv8, predict's input frame needs to change to size(640, 640), so if frame is not resized into a square, bounding box will be deformed and the complete object will not be captured.

File: samples/dnn/object_detection.py (hunk 3 above)
Comment (dkurt) [reply samples/dnn/object_detection.py:100]: `readNet` can handle empty string for config:
```
readNet("model.onnx", "", "onnx")
```

File: samples/dnn/object_detection.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dkurt) [reply samples/dnn/object_detection.py:314]: There are `args.width` and `args.height` for it which are used in `blobFromImage`.

[the remark this answers is not in the document]
Comment (JasonLin1110) [reply samples/dnn/object_detection.py]: Hi, we have removed resizing the frame and used scaling bounding box instead. 
Please check, thanks!

File: samples/dnn/object_detection.py [hunk 6]
```diff
@@ -78,15 +80,17 @@
...
 if args.classes:
     with open(args.classes, 'rt') as f:
         classes = f.read().rstrip('\n').split('\n')
 
 # Load a network
-net = cv.dnn.readNet(cv.samples.findFile(args.model), cv.samples.findFile(args.config), args.framework)
+if args.config != None:
+    net = cv.dnn.readNet(cv.samples.findFile(args.model), cv.samples.findFile(args.config), args.framework)
+else:
+    net = cv.dnn.readNet(cv.samples.findFile(args.model), '', args.framework)
```
Comment (dkurt) [inline samples/dnn/object_detection.py:93]: `findFile` already called before, so let's keep just this:
```python
net = cv.dnn.readNet(args.model, args.config, args.framework)
```

File: samples/dnn/object_detection.py [hunk 7]
```diff
@@ -54,15 +54,17 @@
...
 add_preproc_args(args.zoo, parser, 'object_detection')
 parser = argparse.ArgumentParser(parents=[parser],
                                  description='Use this script to run object detection deep learning networks using OpenCV.',
                                  formatter_class=argparse.ArgumentDefaultsHelpFormatter)
 args = parser.parse_args()
-
 args.model = findFile(args.model)
-args.config = findFile(args.config)
+if args.config != None:
+    args.config = findFile(args.config)
```
Comment (dkurt) [inline samples/dnn/object_detection.py:67]: args.config = findFile(args.config) if args.config is None else ""

File: samples/dnn/object_detection.py [hunk 8]
```diff
@@ -54,15 +54,17 @@
                     dest='asyncN',
                     help='Number of asynchronous forwards at the same time. '
                          'Choose 0 for synchronous mode')
+parser.add_argument('--postprocessing', choices=['yolo', 'ssd', 'faster_rcnn', 'yolov8'],
+                    help='Optional name of AI algo. ')
```
Comment (dkurt) [inline samples/dnn/object_detection.py:58]: Not representative description. What about "Detection results post-processing kind depends on model topology"

File: samples/dnn/object_detection.py [hunk 9]
```diff
@@ -183,6 +205,24 @@ def drawPred(classId, conf, left, top, right, bottom):
...
+        for i in range(len(nms_indices)):
+            index = nms_indices[i]
+            nms_boxes.append(boxes[index])
+            nms_classIds.append(classIds[index])
+            nms_confidences.append(confidences[index])
+            indices.append(i)
+        boxes = nms_boxes
+        classIds = nms_classIds
+        confidences = nms_confidences
+    ###
```
[the remark this answers is not in the document]
Comment (JasonLin1110) [reply samples/dnn/object_detection.py]: Hi , we have modified those, please check, thanks.

File: samples/dnn/object_detection.py (hunk 8 above)
Comment (JasonLin1110) [reply samples/dnn/object_detection.py]: We have changed the description as above, thanks.

**Comment by Tsai-chia-hsiang** [comment]
@dkurt 
Hi, during refactoring the code, we got a question about line 54: 
image

However, according to the comment above it, shoudn't it be ```score = detection[4:]``` ? Since detection[0,1,2,3] are the bbox x,y,w,h and [4: ] are the score. 

Or Do I misunderstand the usage of this line ?  

Thanks

File: samples/dnn/models.yml [hunk 10]
```diff
@@ -63,6 +63,19 @@ tiny-yolo-voc:
...
+  load_info:
+    url: "https://github.com/JasonLin1110/ultralytics_yolov8_onnx_model/yolov8net.onnx"
+    sha1: "67144099b4d8c40fa2d1f8cd17b2a6c6bb81ce90"
+  model: "yolov8net.onnx"
+  mean: [0, 0, 0]
+  scale: 0.00392
+  width: 640
+  height: 640
+  rgb: true
+  classes: "object_detection_classes_yolov8.txt"
```
Comment (dkurt) [inline samples/dnn/models.yml:76]: missed file

File: samples/dnn/models.yml [hunk 11]
```diff
@@ -63,6 +63,19 @@ tiny-yolo-voc:
   classes: "object_detection_classes_pascal_voc.txt"
   sample: "object_detection"
 
+yolov8:
+  load_info:
+    url: "https://github.com/JasonLin1110/ultralytics_yolov8_onnx_model/yolov8net.onnx"
```
Comment (dkurt) [inline samples/dnn/models.yml:68]: broken url

Comment (dkurt) [reply samples/dnn/models.yml:68]: BTW, how model was generated (onnx file)? Is there any officially supported supported sources?

File: samples/dnn/models.yml (hunk 10 above)
Comment (JasonLin1110) [reply samples/dnn/models.yml:76]: here's our classes: "https://github.com/JasonLin1110/ultralytics_yolov8_onnx_model/blob/main/object_detection_classes_yolov8.txt", should we update new url to models.yml?

File: samples/dnn/models.yml (hunk 11 above)
Comment (JasonLin1110) [reply samples/dnn/models.yml:68]: Here's the url: "https://github.com/JasonLin1110/ultralytics_yolov8_onnx_model/blob/main/yolov8net.onnx".
We download yolov8n.pt by ```model = YOLO('yolov8n.pt')``` and use ```model.export('onnx')``` to get yolov8n.onnx, yolov8net.onnx just rename yolov8n.onnx. We didn't find onnx file in official, but use export('onnx') can change pt file to onnx

File: samples/dnn/models.yml (hunk 10 above)
Comment (dkurt) [reply samples/dnn/models.yml:76]: This is identical to https://github.com/opencv/opencv/blob/4.x/samples/data/dnn/object_detection_classes_yolov4.txt

Comment (JasonLin1110) [reply samples/dnn/models.yml]: We have updated url and classes in models.yml's yolov8 part, please check. Thank you!

**Comment by Tsai-chia-hsiang** [comment]
@dkurt 
Hi, we've seen that we didn't the "doc" part of unit test . Is there any mistake we made that cause this error ?

image

Thanks.

**Comment by dkurt** [comment]
@Tsai-chia-hsiang, made some changes with sample update and common parts unification. Please give your comment on the padding preprocessing (I removed it for now). If you don't mind, we can introduce it in the next PR.

Also, I have replaced onnx URL to Ultralytics's `.pt` model with a comment how to convert to ONNX.

Now sample works with the following commands:
YOLOv8 (Ultralytics):
```
python samples/dnn/object_detection.py yolov8 --model ~/yolov8net.onnx --input kodim23.png
```

YOLOv4 (Darknet):
```
python samples/dnn/object_detection.py yolo --model ~/yolov4.weights --input kodim23.png
```

SSD (Caffe):
```
python samples/dnn/object_detection.py ssd_caffe --input kodim23.png
```

**Comment by Tsai-chia-hsiang** [comment]
Ok, should I add the comment about the preprocess part in the code and commit again 
or I just leave the comment hear and you could directly combine it into the code? 

Thanks.

**Review by dkurt — APPROVED** [review]

**Comment by Tsai-chia-hsiang** [comment]
OK, I see.

Thanks!

**Comment by asmorkalov** [comment]
```
 python samples/dnn/object_detection.py yolov8 --model ./samples/yolov8net.onnx --input ./samples/data/lena.jpg 
[ WARN:0@0.020] global samples.cpp:61 findFile cv::samples::findFile('object_detection_classes_yolov4.txt') => ''
Traceback (most recent call last):
  File "samples/dnn/object_detection.py", line 90, in <module>
    net = cv.dnn.readNet(args.model, args.config, args.framework)
cv2.error: OpenCV(4.8.0-dev) /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/onnx/onnx_importer.cpp:279: error: (-210:Unsupported format or combination of formats) Failed to parse ONNX model: ./samples/yolov8net.onnx in function 'ONNXImporter'
```

**Comment by dkurt** [comment]
@asmorkalov , dis you convert a model to ONNX or used https://github.com/JasonLin1110/ultralytics_yolov8_onnx_model/blob/main/yolov8net.onnx ? I tested second option, let me try again.

File: samples/dnn/models.yml [hunk 12]
```diff
@@ -75,6 +78,25 @@ tiny-yolo-voc:
...
+  background_label_id: 0
+  sample: "object_detection"
+
+# Model should be converted to ONNX:
+# from ultralytics import YOLO
+# model = YOLO("yolov8n.pt")
+# model.export(format="onnx")
+yolov8:
+  load_info:
+    url: "https://github.com/ultralytics/assets/releases/download/v0.0.0/yolov8n.pt"
```
Comment (asmorkalov) [inline samples/dnn/models.yml:90]: ONNX link is required here.

**Comment by asmorkalov** [comment]
Ok. I found out the issue reason.
- config for download_models.py contains link to pt, but not onnx. The file is renamed to .onnx as it's defined in yaml config.
- In case if I use ONNX from https://github.com/JasonLin1110/ultralytics_yolov8_onnx_model/blob/main/yolov8net.onnx the sample works fine.

File: samples/dnn/models.yml (hunk 12 above)
Comment (dkurt) [reply samples/dnn/models.yml:90]: It's better to rename `model: "yolov8net.onnx"` to `model: "yolov8n.pt"`.

> config for download_models.py contains link to pt, but not onnx. The file is renamed to .onnx as it's defined in yaml config.

I have not seen entry for YOLOv8 in download_models.py except https://github.com/CVHub520/X-AnyLabeling/releases/download/v0.1.0/yolov8n.onnx which is not an official repo. Can we keep a link to `.pt`?

Comment (dkurt) [reply samples/dnn/models.yml:90]: Oh, excuse me. I just missed that there is another https://github.com/opencv/opencv/blob/4.x/samples/dnn/download_models.py

Comment (dkurt) [reply samples/dnn/models.yml]: Replaced to [URL from opencv_extra](https://github.com/opencv/opencv_extra/blob/0c9ec227253c60165b125c7bce313fc2bb4cb592/testdata/dnn/download_models.py#L1008)

**Review by asmorkalov — APPROVED** [review]
:+1: 
Tested manually with current OpenCV 4.x, Ubuntu 18.04 and ONNX model by link in yaml config.

**Comment by dkurt** [comment]
@asmorkalov, thanks for testing!
